source/manual/architecture-deep-dive.html.md

```diff
@@ -0,0 +1,462 @@
+["Our content delivery network"][our-cdn] for more information). If Fastly
+doesn't have a page in its cache, it fetches the page from origin.
+
+Caches can be purged using the [cache-clearing-service], which tells Fastly to
+soft-purge (i.e. only remove the cached version once it has received the new
+version from origin). This cache clearing service is triggered automatically
+when pages are updated - more on that later.
+
+[cache-clearing-service]: https://github.com/alphagov/cache-clearing-service
+[Fastly]: https://www.fastly.com/
+[our-cdn]: /manual/cdn.html
+[Varnish]: https://varnish-cache.org/
+
+#### Failover
+
+If a Fastly request to origin returns a 5xx response, Fastly will request
+content from the mirror, which is static HTML hosted in an S3 bucket on AWS.
+The contents of the mirror are updated daily via a [govuk-crawler-worker],
+which recursively crawls GOV.UK URLs from a message queue, visiting the pages
+and saving the output to disk.
+
+[govuk-crawler-worker]: https://github.com/alphagov/govuk_crawler_worker
+
+#### Routing on the CDN
+
+As well as for caching, Varnish is used for the redirection from `gov.uk` to
+`www.gov.uk`, which is configured in Varnish Configuration Language (VCL) and
+uploaded directly to Fastly via [govuk-cdn-config].
+
+Other redirects that happen at the Fastly level include [bouncer]: a GOV.UK
+application responsible for redirecting traffic from old pre-GOV.UK websites.
+This is configured via [transition] and [transition-config].
+
+[bouncer]: https://github.com/alphagov/bouncer
+[govuk-cdn-config]: https://github.com/alphagov/govuk-cdn-config
+[transition]: https://github.com/alphagov/transition
+[transition-config]: https://github.com/alphagov/transition-config
+
+### Routing on GOV.UK
+
+#### Getting to the 'router' application
+
+![](images/load-balancer.png)
+
+Some requests make it through the CDN and cache layers to 'origin'. Origin is
+a stack of computers in the cloud - in this case, AWS - and its entry point is
+a load balancer.
+
+The load balancer knows based on the hostname which machine 'class' to route
+to. Different classes of machine run different sets of GOV.UK applications.
+How many machines are allocated to a class - and how big those machines are -
+is configured using [Terraform], in [govuk-aws]. What runs on each machine
+class is configured in govuk-puppet, a file and process management system we'll
+cover in more detail later.
+
+![](images/nginx-routing.png)
+
+External requests are routed to a 'cache' machine, which uses Varnish for its
+caching. If Varnish has the route response in its cache, it returns that,
+otherwise it forwards the request to [Nginx], a web server running on the
+machine. Nginx is configured with govuk-puppet.
+
+Nginx proxies most requests to [router], which is a GOV.UK maintained
+application running on the cache machines. However, Nginx does proxy some
+routes directly to other apps, such as image URLs routing to asset-manager.
```

Comment:
I believe the requests are processed in this order on the cache machine: NGinx -> Varnish -> Router (at least that's what it looks like if you look at the NGinx and Varnish configuration).

Reply:
That makes more sense to me. I'll need to update my slides too 👍